src/Analyzers/Core/Analyzers/PopulateSwitch/PopulateSwitchExpressionHelpers.cs

```diff
@@ -103,4 +102,36 @@ public static bool IsDefault(IPatternOperation pattern)
             },
             _ => false
         };
+
+    public static bool HasBothNullAndUnderlyingValueCases(ISwitchExpressionOperation operation)
+    {
+        var type = operation.Value.Type;
+        var underlyingType = type.IsNullable(out var underlying) ? underlying : type;
+
+        var hasNullArm = false;
+        var hasUnderlyingTypeArm = false;
+
+        foreach (var arm in operation.Arms)
+        {
+            var pattern = arm.Pattern;
+
+            if (pattern is IConstantPatternOperation { Value: IConversionOperation { ConstantValue: { HasValue: true, Value: null } } })
+            {
+                hasNullArm = true;
+                continue;
+            }
+
+            var matchedType = pattern switch
+            {
+                ITypePatternOperation typePattern => typePattern.MatchedType,
+                IDeclarationPatternOperation declarationPattern => declarationPattern.MatchedType,
+                _ => null
+            };
+
+            if (SymbolEqualityComparer.Default.Equals(matchedType, underlyingType))
+                hasUnderlyingTypeArm = true;
+        }
+
+        return hasNullArm && hasUnderlyingTypeArm;
```

Comment:
Put in for loop so it can return immediately?

Reply:
Didn't understand this one. We can only be sure we have both arms/cases only if we loop through all of them and check each. How can that be short-circuited?

Reply:
I pushed a change where i moved this into the loop. ;)